File: ServerSide/BallPassSchedulePattern/Players_AiTypes/EmbeddingPlayerDir/EmbeddingLayer.py
import numpy as np

class EmbeddingLayer:
    def __init__(self, initial_weights):
        """
        初期化
        :param initial_weights: 単語埋め込み行列（2D配列 or NumPy配列）
        """
        self.weights = np.array(initial_weights)  # 埋め込み行列
        self.gradients = np.zeros_like(self.weights)  # 勾配行列を初期化
        self.current_token_id = None  # 現在のトークンID
        self.current_dx = None  # Backwardで計算されるdx

    def forward(self, token_id):
        """
        トークンIDに対応する埋め込みベクトルを返す
        :param token_id: int, トークンID
        :return: 単語埋め込みベクトル（1D NumPy配列）
        """
        if token_id < 0 or token_id >= self.weights.shape[0]:
            raise ValueError("Invalid token_id or weights not initialized.")
        
        self.current_token_id = token_id
        return self.weights[token_id]

    def backward(self, gradient):
        """
        勾配を更新し、dxを計算する
        :param gradient: 1D NumPy配列, トークンIDに対応する埋め込みベクトルの勾配
        :return: dx (1D NumPy配列)
        """
        if self.current_token_id is None:
            raise ValueError("No token_id has been processed in forward.")
        
        if gradient.shape != self.weights[self.current_token_id].shape:
            raise ValueError("Gradient shape mismatch.")

        # 勾配を蓄積
        self.gradients[self.current_token_id] += gradient
        self.current_dx = gradient

        return self.current_dx

    def get_gradients(self):
        """
        蓄積された勾配を返す
        :return: 勾配行列（2D NumPy配列）
        """
        return self.gradients

    def clear_gradients(self):
        """
        勾配をクリア
        """
        self.gradients.fill(0)  # NumPyの機能で全てを0にする

File: InstanceSide/BallPassSchedulePattern/Players_AiTypes/EmbeddingPlayerDir/EmbeddingLayer.cs
using UdonSharp;
using UnityEngine;


public class EmbeddingLayer : UdonSharpBehaviour
{
    // クラス内で保持する変数
    public RinaNumpy rinaNumpy;

    private float[][] weights; // 単語表現ベクトル（埋め込み行列）
    private float[][] gradients; // 勾配を保持する配列
    private int currentTokenId; // Forwardで保持する現在のトークンID
    private float[] currentDx; // Backwardで保持する現在のトークンのdx


    public float[] Forward(int tokenId, float[][] W)
    {
        // 重みのアタッチ
        weights = W

        // 入力トークンIDを保持
        currentTokenId = tokenId;

        return weights[tokenId];
    }

    public float[] Backward(float[] gradient, int TokenId, float[][] dW)
    {
        // パラメータの参照
        gradients　= dW

        // 勾配を蓄積
        for (int i = 0; i < gradients[TokenId].Length; i++)
        {
            gradients[TokenId][i] += gradient[i];
        }

        // dxを保持
        currentDx = gradient;

        // dxを返す
        return currentDx;
    }

    public float[][] GetGradients()
    {
        // 勾配を取得（他のプロセスで更新するため）
        return gradients;
    }

    public void ClearGradients()
    {
        // 勾配をクリア（次の更新に備える）
        for (int i = 0; i < gradients.Length; i++)
        {
            for (int j = 0; j < gradients[i].Length; j++)
            {
                gradients[i][j] = 0f;
            }
        }
    }
}
